### PR TITLE
add version in setup.cfg

### PR DESCRIPTION
**Pull request by jackeichen**
add version in setup.cfg, so we can get the version info with pip tool, which make it more convenience  for user use the scsi lib.

### CODE REVIEW

**Review by rosjat — CHANGES_REQUESTED** [review]
@jackeichen  first of all thanks for the contribution but i saw a few things you should fix before the pull request.

 - in your cdb file dont use an import that no one else wil have (from your project), simply import the DSCSIDevice from pyscsi
 - you might wanna edit the copyright notice since you did the work and i dont know if ronnie wants to take the credit away ;)
 - helper functions like lba_convert might be better of in utils and if its not a general convert try to name it more explicit (Example scsi_to_ata_lba_convert)
 - also in scsi_device.py there is no way to pass sense data to the execute method, the only thing you can pass (and its a default atm, the max lenght of sense data you expect to get). So you need to issue the cdb to the device and then check if you got sense data or not (@sahlberg  can be more specific here i guess)
 - in the scsi_sense I dont think you need extra stuff, as far as i see the sense data you get back is in fixed format

**Comment by jackeichen** [comment]
@rosjat  Sorry i made some mistakes: the import error(I should test the code in a fresh environment), and used the old cython-sgio(So i suggest to give the version info to the setuptools).

The last two questions:
@sahlberg  I usually cannot get the cmd sense data like this(cmd.sense is None):
    with SCSI(device) as s:
        print('ReadCapacity10')
        cmd = s.readcapacity10()
        print (cmd.sense)
        r = cmd.result
        for k, v in r.items():
            print('%s - %s' % (k, v))

If any other interface to get cmd.sense, please let me know. I need it to decode ATA Return Descriptor. 

@rosjat  After execute ata-passthroug, the ATA Return Descriptor is included in cmd sense data[8:22], the response_code is 0x72h, and I remember that the first 8 bytes sense data[0:8] is in SCSICheckCondition._desc_format_sdata_bits format(long ago to read the spec, i'm not sure now, but the ATA Return Descriptor is in sense data[8:22]).  Should i get a wrong decode result If use fixed format?

Here is the ATA Return Descriptor, used to show ATA command output, offset=8:
ata_status_return_descriptor=   {'descriptor_code': [0xff, 0],
                                                   'additional_descriptor_length': [0xff, 1],
                                                   'extend': [0x01, 2],
                                                   'error': [0xff, 3],
                                                   'sector_count_rsvd': [0xff, 4],
                                                   'sector_count': [0xff, 5],
                                                   'lba_low_rsvd': [0xff, 6],
                                                   'lba_low': [0xff, 7],
                                                   'lba_mid_rsvd': [0xff, 8],
                                                   'lba_mid': [0xff, 9],
                                                   'lba_high_rsvd': [0xff, 10],
                                                   'lba_high': [0xff, 11],
                                                   'device': [0xff, 12],
                                                   'status': [0xff, 13], }

**Comment by rosjat** [comment]
> @rosjat After execute ata-passthroug, the ATA Return Descriptor is included in cmd sense data[8:22], the response_code is 0x72h, and I remember that the first 8 bytes sense data[0:8] is in SCSICheckCondition._desc_format_sdata_bits format(long ago to read the spec, i'm not sure now, but the ATA Return Descriptor is in sense data[8:22]). Should i get a wrong decode result If use fixed format?

well the doc states that :
```

12.2.2.8 Fixed format sense data
Table 211 shows the fields returned in the fixed format sense data (see SPC-5) for ATA PASS-THROUGH
commands.
```

so that would mean you will get fixed formated data that you can then unmarshall

**Comment by jackeichen** [comment]
Thanks for the detailedly review, i think i can decode the cmd sense, I need get the cmd sese first, anyone can help?

**Comment by sahlberg** [comment]
> @sahlberg I usually cannot get the cmd sense data like this(cmd.sense is None): with SCSI(device) as s: print('ReadCapacity10') cmd = s.readcapacity10() print (cmd.sense) r = cmd.result for k, v in r.items(): print('%s - %s' % (k, v))
> 
> If any other interface to get cmd.sense, please let me know. I need it to decode ATA Return Descriptor.

Right now we only return it in .sense IFF the command failed with check-condition.
I struggle to find a copy of the actual document so I can only give some best effort response.

Right now, we only return sense from sgio IFF the command failed with check condition.
Do you mean that for SAT, that we get sense also for other conditions?
I.e. do we get sense data also when the command has not failed with check condition?

If so, there are two things you can try.
Around this part of the code https://github.com/python-scsi/cython-sgio/blob/be246e6f1f994bfbe2bdcecaf275e9007f696a77/src/linux_sgio.pyx#L178

1, you could check the scsi OP code and unconditionally try to return the content of the sense buffer by raising an exception if it is one of the passthrough opcodes.
2, If you can look at the content of the sense buffer and identify it is SAT related, then again you can raise an exception and pass the sense buffer.

**Comment by sahlberg** [comment]
> 



> Thanks for the detailedly review, i think i can decode the cmd sense now after reading the spec, I need get the cmd sense first, anyone can help?
> 
> @rosjat @sahlberg I cannot find any interface to get the command sense for now after reading the project python-scsi and cython-sgio. It seems that the sense data init and check in cython-sgio, it doesn't provide a api to send it out unless check sense error. I think it’s reasonable to pass sense data to user as a scsi library.

Yes, that is correct. Initially it was assumed you would never need the sense buffer except if it was a check-condition so the api is missing.
There are a couple of ways to handle this. Right now execute() just returns a scalar that is the number of bytes sent/received.
You could change this to be a tuple instead and return (io_hdr.resid, bytes(sense[:io_hdr.sb_len_wr]))
There should only be a few places in the code where we call execute  so would only need to be changed in a small number of places.

**Comment by jackeichen** [comment]
> 

@rosjat  Yes, this is what I need. 

In ATA command set, I usually need the cmd sense data, in which ATA command out(ATA command Error field, Status filed, LBA field) is included. 
For example, I send a AccessibleMaxAddressCfg command, then the address will be stored in the sense data->LBA field, but it cannot be got if command success.

In additional, there is a variable scsi_command.SCSICommand._sense, it exists but not used  in the code(always None now), so i think:
1. In cython-sgio, change this to be two values instead and "return io_hdr.resid, bytes(sense[:io_hdr.sb_len_wr])"；
2. In scsi_device.py, change SCSIDevice.execute to:
   `    try:
            resid,cmd.sense = sgio.execute(
                                                self._file,
                                                cmd.cdb,
                                                cmd.dataout,
                                                cmd.datain)
        except sgio.CheckConditionError as error:
            self.CheckCondition(error.sense)
            # if it need here to pass it to user?
            # cmd.sense = error.sense
`

**Comment by rosjat** [comment]
> 1. In cython-sgio, change this to be two values instead and "return io_hdr.resid, bytes(sense[:io_hdr.sb_len_wr])"；
> 2. In scsi_device.py, change SCSIDevice.execute to:
>    `resid,cmd.sense = sgio.execute(self._file, cmd.cdb,cmd.dataout,cmd.datain) `
@jackeichen  as @sahlberg stated you can try to modify the code so it does that. I will take a look at the sgio code but if you feel convident simply modify and make a pull request there @Flameeyes  can for sure review it

**Comment by jackeichen** [comment]
> > @rosjat After execute ata-passthroug, the ATA Return Descriptor is included in cmd sense data[8:22], the response_code is 0x72h, and I remember that the first 8 bytes sense data[0:8] is in SCSICheckCondition._desc_format_sdata_bits format(long ago to read the spec, i'm not sure now, but the ATA Return Descriptor is in sense data[8:22]). Should i get a wrong decode result If use fixed format?
> 
> well the doc states that :
> 
> ```
> 
> 12.2.2.8 Fixed format sense data
> Table 211 shows the fields returned in the fixed format sense data (see SPC-5) for ATA PASS-THROUGH
> commands.
> ```
> 
> so that would mean you will get fixed formated data that you can then unmarshall

This is what i get from the spec and test about the sense data:

1. If i set ck_cond=0 when build ata-passthrough command, then no sense data returned;
2. if i set ck_cond=1, then ResponseCode=0x72h (Sense data format Descriptor, SPC-4 chapter 4.5 sense data) returned, i guess that scsi_sense._desc_format_sdata_bits is used for decoding this, correct me if not. for ata-passthrough sense data[8:]->Sense data descriptor is equal to ATA Return Descriptor;
3. Set protocal=0x0f(and no matter ck_cond=0 or 1) when build ata-passthrough command, then no actual ata command is executed, this will return the contents of the ATA Command Block in the ATA Status Return Descriptor, can be decoded by 12.2.2.8 Fixed format sense data

I think no need to execute two times in one command, just set ck_cond=1 is more convenient.

**Comment by rosjat** [comment]
@jackeichen  if ck_cond is set to 1 you will get what you want, it will return sense data no matter what and then you have to unmarshall the data so you can make sense of it. If ck_cond is set to 0 it will only return sense data is there was an error and then you get a sense data with a sense code to determine what went wrong.

**Comment by jackeichen** [comment]
> @jackeichen if ck_cond is set to 1 you will get what you want, it will return sense data no matter what and then you have to unmarshall the data so you can make sense of it. If ck_cond is set to 0 it will only return sense data is there was an error and then you get a sense data with a sense code to determine what went wrong.
> 
> so with ck_cond = 1 you either get data you need or a checkcondition to handle an error

Totally right, a few of ata commands need LBA field or counts field in sense data to get the command out, all the ata commands need the  sense data to decide what went wrong when error, like devide not ready,media error, or something else.

**Review by Flameeyes — CHANGES_REQUESTED** [review]
Changing `setup.cfg` should not be necessary. `pip` can detect the version just fine with `setup.py`.

Actually, fixing the version there means that the tag-and-release workflow will break, so please drop that commit from the pull request.

**Comment by jackeichen** [comment]
> Changing `setup.cfg` should not be necessary. `pip` can detect the version just fine with `setup.py`.
> 
> Actually, fixing the version there means that the tag-and-release workflow will break, so please drop that commit from the pull request.

I will drop it

**Comment by jackeichen** [comment]
@sahlberg @rosjat  I have add the raw_sense_data,  and test some command(Non-data, PIO Data-in, PIO data-out),  it works well.
Waiting the code change in cython-sgio.

**Comment by sahlberg** [comment]
> @sahlberg @rosjat I have add the raw_sense_data, and test some command(Non-data, PIO Data-in, PIO data-out), it works well. Waiting the code change in cython-sgio.

I just created a pull request for cython-sgio. Please have a look: https://github.com/python-scsi/cython-sgio/pull/8
In order to get the sense buffer you need to call execute() with the new named argument return_sense_buffer=True
I did it that way so that cython-sgio will still be compatible with older python-scsi that will not yet request the sense buffer.

**Comment by jackeichen** [comment]
> > @sahlberg @rosjat I have add the raw_sense_data, and test some command(Non-data, PIO Data-in, PIO data-out), it works well. Waiting the code change in cython-sgio.
> 
> I just created a pull request for cython-sgio. Please have a look: [python-scsi/cython-sgio#8](https://github.com/python-scsi/cython-sgio/pull/8) In order to get the sense buffer you need to call execute() with the new named argument return_sense_buffer=True I did it that way so that cython-sgio will still be compatible with older python-scsi that will not yet request the sense buffer.

Then could you help to add the interface to scsi_device.execute and scsi.execute? Usually more code from maintaner, more reasonable.

**Comment by sahlberg** [comment]
> > > @sahlberg @rosjat I have add the raw_sense_data, and test some command(Non-data, PIO Data-in, PIO data-out), it works well. Waiting the code change in cython-sgio.
> > 
> > 
> > I just created a pull request for cython-sgio. Please have a look: [python-scsi/cython-sgio#8](https://github.com/python-scsi/cython-sgio/pull/8) In order to get the sense buffer you need to call execute() with the new named argument return_sense_buffer=True I did it that way so that cython-sgio will still be compatible with older python-scsi that will not yet request the sense buffer.
> 
> Then could you help to add the interface to scsi_device.execute and scsi.execute? Usually more code from maintaner, more reasonable.

I can do that once the patch to the cython package is merged.

**Review by sahlberg — COMMENTED** [review]
We should also have a test for the marshalling/unmarshalling of cdb and data.

File: pyscsi/pyscsi/scsi_cdb_atapassthrough12.py [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+    result += ((lba >> 16) & 0xFF)
+    return result
+
+
+class ATAPassThrough12(SCSICommand):
+    """
+    A class to send a ATAPassThrough12 command to a ata device
+    """
+    _cdb_bits = {'opcode': [0xff, 0],
+                 'reserved1':[0x01,1],
```
[the remark this answers is not in the document]
Comment (jackeichen) [reply pyscsi/pyscsi/scsi_cdb_atapassthrough12.py]: I have made some change to the code and commit: 
1. drop "reserved" and "Obsolete" by @sahlberg ; 
2. drop version in setup.cfg @Flameeyes ;
3. fix the mistake by @rosjat ;
4. keep original code as much as possible; 

Let me know if any other suggestion.

**Review by rosjat — APPROVED** [review]

**Comment by rosjat** [comment]
@jackeichen  code sarting to look solid now and if some achnages arise we can do work on it when mwe pulled the changes.

@sahlberg  if you ok to i would merge the code alter and we can work on some tests too

**Comment by sahlberg** [comment]
I am fine with it and I clicked the button.
We need the change to cython to go in too, and some additional trivial patches for the new cython api.
We need @jackeichen input here since he is the subjec tmatter expert on passthrough ata.

@jackeichen what else do we need to do to get your code running?

**Comment by rosjat** [comment]
i msg @Flameeyes  to just check the cython-sgio stuff since the changes shouldnt break anything there but some automation stuff didnt go through. So i guess we should be good there too soon

**Comment by jackeichen** [comment]
> I am fine with it and I clicked the button. We need the change to cython to go in too, and some additional trivial patches for the new cython api. We need @jackeichen input here since he is the subjec tmatter expert on passthrough ata.
> 
> @jackeichen what else do we need to do to get your code running?

I think we need an argument to enable return_sense_buffer:
scsi.atapassthrough12(*args, **kwargs) -> scsi.execute(cmd) -> scsi_device.execute(cmd) -> linux_sgio(*args, **kwargs, return_sense_buffer=True)
scsi.atapassthrough16(*args, **kwargs) -> scsi.execute(cmd) -> scsi_device.execute(cmd) -> linux_sgio(*args, **kwargs, return_sense_buffer=True)

For example:
scsi.atapassthrough12(*args, **kwargs) -> scsi.execute(cmd, raw_sense=True) -> scsi_device.execute(cmd, raw_sense=raw_sense) -> linux_sgio(*args, **kwargs, return_sense_buffer=raw_sense)
and raw_sense=False in default.

That's all.

**Pull request merged**